tests/vm_test.go

```diff
@@ -109,6 +109,47 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
 			Expect(len(reviewResponse.Details.Causes)).To(Equal(1))
 			Expect(reviewResponse.Details.Causes[0].Field).To(Equal("spec.template.spec.domain.devices.disks[2].name"))
 		})
+		It("[rfe_id:4156][case_id:4190][crit:medium][vendor:cnv-qe@redhat.com][level:component] should be rejected when disk is backed by PVC created by DataVolume instead of the DataVolume itself", func() {
+			dv := tests.NewRandomDataVolumeWithHttpImport(tests.GetUrl(tests.AlpineHttpUrl), tests.NamespaceTestDefault, k8sv1.ReadWriteOnce)
+			_, err := virtClient.CdiClient().CdiV1alpha1().DataVolumes(dv.Namespace).Create(dv)
+			Expect(err).To(BeNil())
+
+			defer func(dv *cdiv1.DataVolume) {
+				By("Deleting the DataVolume")
+				ExpectWithOffset(1, virtClient.CdiClient().CdiV1alpha1().DataVolumes(dv.Namespace).Delete(dv.Name, &metav1.DeleteOptions{})).To(Succeed())
+			}(dv)
+			tests.WaitForSuccessfulDataVolumeImport(dv, 60)
+
+			vmi := tests.NewRandomVMI()
+
+			vmi.Spec.Domain.Resources.Requests[k8sv1.ResourceMemory] = resource.MustParse("64M")
+
+			diskName := "disk0"
+			bus := "virtio"
+			vmi.Spec.Domain.Devices.Disks = append(vmi.Spec.Domain.Devices.Disks, v1.Disk{
+				Name: diskName,
+				DiskDevice: v1.DiskDevice{
+					Disk: &v1.DiskTarget{
+						Bus: bus,
+					},
+				},
+			})
+			vmi.Spec.Volumes = append(vmi.Spec.Volumes, v1.Volume{
+				Name: diskName,
+				VolumeSource: v1.VolumeSource{
+					PersistentVolumeClaim: &k8sv1.PersistentVolumeClaimVolumeSource{
+						ClaimName: dv.ObjectMeta.Name,
+					},
+				},
+			})
+
+			vmi.Spec.Domain.Resources.Requests[k8sv1.ResourceMemory] = resource.MustParse("512M")
+
+			vm := tests.NewRandomVirtualMachine(vmi, true)
+			vm.Spec.DataVolumeTemplates = append(vm.Spec.DataVolumeTemplates, *dv)
+			_, err = virtClient.VirtualMachine(tests.NamespaceTestDefault).Create(vm)
```

Comment:
fyi, this is testing the validation webhook, and not the new logic you introduced.

If you want to test your new logic, you'd need to remove the DataVolumeTemplate entry and just attempt to use the PVC directly.

The result right now is going to be that the VMI hangs indefinitely reporting that it can't be scheduled due to the PVC conflict. While this prevents the issue from occurring, it's unclear if this is the best user experience though.

For example, how would the user receive feedback that this issue has occurred?

Reply:
Yes, the test only added that particular case to be verified (i.e. that DV Template and VM Template are consistent).

I've added another test for the case where there's no DV Template but still the PVC is owned by a DV.

Also, I made VMI fail in this case (instead of being in the Pending state for ever).